unittests/test-contracts/savanna/finality_violation/finality_violation.cpp

```diff
@@ -0,0 +1,136 @@
+    auto f_itr = std::find(reversible_blocks_digests.begin(), reversible_blocks_digests.end(), computed_digest);
+
+    check(f_itr==reversible_blocks_digests.end(), "finality digest of low block exists in reversible_blocks_digests vector");
+
+    //Proof of rule #2 finality violation
+
+    savanna::bitset proof_1_bitset(finalizer_policy.finalizers.size(), high_proof.active_policy_qc.finalizers);
+    savanna::bitset proof_2_bitset(finalizer_policy.finalizers.size(), low_proof.active_policy_qc.finalizers);
+
+    auto result = bitset::compare(proof_1_bitset, proof_2_bitset);
+
+    return {result.first.to_string(), result.second.to_string()};
+
+}
+
+//Rule #3 : Do not vote on a block that conflicts with another block on which you are locked
+std::pair<std::string, std::string> finality_violation::rule3(   const finalizer_policy_input& finalizer_policy, 
+                                    const finality_proof& high_proof,
+                                    const finality_proof& low_proof,
+                                    const std::vector<checksum256>& reversible_blocks_digests){
+
+    //Verify QCs
+    check_qcs(finalizer_policy, high_proof, low_proof);
+
+    //Compare timestamps
+    block_timestamp low_proof_timestamp = low_proof.qc_block.level_3_commitments.value().timestamp;
+    block_timestamp high_proof_timestamp = high_proof.qc_block.level_3_commitments.value().timestamp;
+    block_timestamp low_proof_last_claim_timestamp = low_proof.qc_block.level_3_commitments.value().latest_qc_claim_timestamp;
+    block_timestamp high_proof_last_claim_timestamp = high_proof.qc_block.level_3_commitments.value().latest_qc_claim_timestamp;
+
+    //If the high proof timestamp is higher than the low proof timestamp, but the high proof last QC claim timestamp is lower than the low proof last QC claim, the lock was violated
+    bool lock_violation = high_proof_timestamp > low_proof_timestamp && high_proof_last_claim_timestamp < low_proof_timestamp;
+    
+    check(lock_violation, "proofs must demonstrate a lock violation");
+
+    //Compute the merkle root of the reversible digests, and verify that it matches the commitment of the high proof
+    check(get_merkle_root(reversible_blocks_digests) == high_proof.qc_block.level_3_commitments->reversible_blocks_mroot, "reversible_blocks_digests merkle root does not match reversible_blocks_mroot");
+
+    //Compute the finality digest of the low proof
+    checksum256 computed_digest = block_finality_data_internal(low_proof.qc_block).finality_digest();
+
+    //Verify that the computed digest for the low proof doesn't appear in the list of reversible block digests committed to by the high proof
+    auto f_itr = std::find(reversible_blocks_digests.begin(), reversible_blocks_digests.end(), computed_digest);
+
+    check(f_itr==reversible_blocks_digests.end(), "finality digest of low block exists in reversible_blocks_digests vector");
```

Comment:
Again, this check is not correct for the same reasons described earlier for the rule 2 proof action. But in addition to that it is attempting to show that the low proof block is not an ancestor of the high proof block, but rule 3 requires that the block claimed by the low proof block is not an ancestor of the high proof block.

A similar process as before needs to be carried out here:

The proper way to check this is as follows:

First, the user must submit the contents of the leaf node in the reversible block Merkle tree (along with the proof of inclusion up to the reversible blocks root) for the block that has a `timestamp` greater than or equal to `low_proof_last_claim_timestamp` and a `parent_timestamp` strictly less than `low_proof_last_claim_timestamp`. Both of those inequalities must be validated otherwise the submitted proof is not correct.

Then, if the `timestamp` for the submitted reversible blocks leaf node is strictly greater than `low_proof_last_claim_timestamp`, then you know that the block claimed by the low proof block is not an ancestor of the high proof block and so a rule 3 violation has occurred.

Otherwise, in the case where the `timestamp` for the submitted reversible blocks leaf node is exactly equal to `low_proof_last_claim_timestamp`, you then need to get finality digest of the block claimed by the low proof block (`low_proof.qc_block.level_3_commitments->latest_qc_claim_finality_digest`) and compare it to the `finality_digest` of the submitted reversible blocks leaf node to check that they are not the same. If they are the same, then again the submitted proof is not correct. But if they are not the same, then you know that the block claimed by the low proof block is not an ancestor of the high proof block and so a rule 3 violation has occurred.

Reply:
Addressed in c7ade593c92904e40a4765608fc7294e133ef006